src/theme/mod.rs

```diff
@@ -205,16 +207,86 @@ where
 
 #[derive(PartialEq, Debug, Default)]
 struct ExtensionMappings {
-    mappings: Vec<(glob::Pattern, Style)>,
+    mappings: Vec<GlobPattern>,
+}
+
+#[derive(PartialEq, Debug)]
+enum GlobPattern {
+    Complex(glob::Pattern, Style),
+
+    // TODO: the String here should probably be an inline string, since we're likely working with
+    // small strings (file extensions).
+    Simple(HashMap<String, Style>),
 }
 
 impl ExtensionMappings {
+    // helper for test suite
+    #[cfg(test)]
+    fn to_vec_pat_style(&self) -> Vec<(glob::Pattern, Style)> {
+        let mut out = Vec::new();
+        for map in &self.mappings {
+            match map {
+                GlobPattern::Complex(p, s) => {
+                    out.push((p.clone(), *s));
+                }
+                GlobPattern::Simple(h) => {
+                    let mut simple_pats = h
+                        .iter()
+                        .map(|(k, v)| (glob::Pattern::new(&format!("*.{}", k)).unwrap(), *v))
+                        .collect::<Vec<(glob::Pattern, Style)>>();
+
+                    simple_pats.sort_by_key(|x| x.0.clone());
+
+                    out.extend(simple_pats);
+                }
+            }
+        }
+        out
+    }
+
     fn is_non_empty(&self) -> bool {
         !self.mappings.is_empty()
     }
 
     fn add(&mut self, pattern: glob::Pattern, style: Style) {
-        self.mappings.push((pattern, style));
+        match self.mappings.last_mut() {
+            None => match is_simple_pattern(pattern) {
+                Ok(s) => {
+                    self.mappings
+                        .push(GlobPattern::Simple(HashMap::from([(s, style)])));
+                }
+                Err(p) => {
+                    self.mappings.push(GlobPattern::Complex(p, style));
+                }
+            },
+            Some(gp) => match (gp, is_simple_pattern(pattern)) {
+                (GlobPattern::Simple(h), Ok(s)) => {
+                    h.insert(s, style);
+                }
+                (_, Ok(s)) => {
+                    self.mappings
+                        .push(GlobPattern::Simple(HashMap::from([(s, style)])));
+                }
+                (_, Err(p)) => {
+                    self.mappings.push(GlobPattern::Complex(p, style));
+                }
+            },
+        }
+    }
+}
+
+fn is_simple_pattern(pattern: glob::Pattern) -> Result<String, glob::Pattern> {
+    if let Some(ext) = pattern.as_str().strip_prefix("*.") {
+        // Maybe too pessimistic here, some of these might be valid.
+        // For example, '*.[*]' is a simple pattern.
+        // Ideally we'd inspect pattern.tokens, but it's not public.
+        if ext.contains(&['?', '*', '[', ']', '.']) {
+            Err(pattern)
+        } else {
+            Ok(ext.to_string())
+        }
+    } else {
```

Comment:
Actually... a file name without an extension doesn't need to be tested against any `GlobPattern::Simple` since it can never match.

Reply:
Actually that doesn't matter